test/ffmpeg-vaapi/transcode/transcoder.py

```diff
@@ -10,35 +10,135 @@
-        "  -hwaccel_output_format vaapi -i {source} -an -c:v {mcodec}"
-        " -vframes {frames} -y {decoded}".format(**vars(self)))
-    elif  vars(self).get("mode", None) == 'hwsw':
-      self.output = call(
-        "ffmpeg -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -v verbose"
-        "  -hwaccel_output_format vaapi -i {source}  -vf 'hwdownload,format=nv12' -an -c:v {mcodec}"
-        " -vframes {frames} -y {decoded}".format(**vars(self)))
-    elif  vars(self).get("mode", None) == 'swhw':
-      self.output = call(
-        "ffmpeg -vaapi_device /dev/dri/renderD128 -v verbose "
-        "-i {source}  -vf 'format=nv12,hwupload' -an -c:v {mcodec}"
-        " -vframes {frames} -y {decoded}".format(**vars(self)))
+  def validate_spec(self):
+    from slash.utils.pattern_matching import Matcher
+
+    assert len(self.outputs), "Invalid test case specification, outputs data empty"
+    assert self.mode in ["sw", "hw"], "Invalid test case specification as mode type not valid"
+
+    # generate platform list based on test runtime parameters
+    iplats, ireq, _ =  self.get_requirements_data("decode", self.codec, self.mode)
+    platforms = set(iplats)
+    requires = [ireq,]
+
+    for output in self.outputs:
+      codec = output["codec"]
+      mode  = output["mode"]
+      assert mode in ["sw", "hw"], "Invalid test case specification as output mode type not valid"
+      oplats, oreq, _ = self.get_requirements_data("encode", codec, mode)
+      platforms &= set(oplats)
+      requires.append(oreq)
+
+    # create matchers based on command-line filters
+    matchers = [Matcher(s) for s in slash.config.root.run.filter_strings]
+
+    # check if user supplied any platform tag on command line
+    pmatch = any(map(lambda p: any([m.matches(p) for m in matchers]), ALL_PLATFORMS))
+
+    # if user supplied a platform tag, check if this test can support it via the
+    # test param-based required platforms
+    if pmatch and not any(map(lambda p: any([m.matches(p) for m in matchers]), platforms)):
+      slash.skip_test("unsupported platform")
+
+    # check required
+    if not all([t for t,m in requires]):
+      slash.skip_test("One or more software requirements not met")
+
+  def gen_input_opts(self):
+    opts = ""
+
+    if "hw" == self.mode:
+      opts += " -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -hwaccel_output_format vaapi"
     else:
-        assert "non supported transcoding"
+      opts += " -vaapi_device /dev/dri/renderD128"
 
-    self.check_output()
-    self.convert_toyuv()
-    self.check_metrics()
+    opts += " -i {source}"
+
+    return opts.format(**vars(self))
+
+  def gen_output_opts(self):
+    self.goutputs = dict()
+
+    opts = "-an"
+
+    for n, output in enumerate(self.outputs):
+      codec = output["codec"]
+      mode = output["mode"]
+
+      _, _, ffparms = self.get_requirements_data("encode", codec, mode)
+      assert ffparms is not None, "gen_output_opts encode parameters are empty"
```

Comment:
How about, "failed to find a suitable encoder"?

Reply:
agreed and modified it